File: detection/signatures.py
from detection.rules import Rule


def default_signatures():
    return [
        Rule(id=1, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="TCP SYN Scan detected", flags="S", severity=2, sid=1000001),

        Rule(id=2, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="TCP NULL Scan detected", flags="0", severity=2, sid=1000002),

        Rule(id=3, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="TCP XMAS Scan detected", flags="FUP", severity=2, sid=1000003),

        Rule(id=4, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="22",
             msg="Possible SSH brute force", severity=3, sid=1000004),

        Rule(id=5, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="21",
             msg="FTP connection attempt", severity=1, sid=1000005),

        Rule(id=6, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="23",
             msg="Telnet connection (cleartext protocol)", severity=2, sid=1000006),

        Rule(id=7, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="SQL injection attempt (UNION SELECT)",
             content="UNION SELECT", nocase=True, severity=3, sid=1000007),

        Rule(id=8, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="SQL injection attempt (OR 1=1)",
             content="OR 1=1", nocase=True, severity=3, sid=1000008),

        Rule(id=9, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="XSS attempt (<script>)",
             content="<script>", nocase=True, severity=3, sid=1000009),

        Rule(id=10, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="Shell injection attempt (/bin/sh)",
             content="/bin/sh", severity=3, sid=1000010),

        Rule(id=11, action="alert", protocol="ICMP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="ICMP packet (potential flood)", severity=1, sid=1000011),

        Rule(id=12, action="alert", protocol="UDP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="53",
             msg="DNS TXT query (possible DNS tunneling)", severity=2, sid=1000012),

        Rule(id=13, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="445",
             msg="SMB access (potential EternalBlue/ransomware)", severity=2, sid=1000013),

        Rule(id=14, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="3389",
             msg="RDP connection attempt", severity=2, sid=1000014),

        Rule(id=15, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="Possible reverse shell (bash -i)",
             content="bash -i", nocase=True, severity=3, sid=1000015),

        Rule(id=16, action="alert", protocol="TCP", src_ip="any", src_port="any",
             dst_ip="any", dst_port="any",
             msg="Possible reverse shell (nc -e)",
             content="nc -e", severity=3, sid=1000016),
    ]

File: detection/signatures.go
package detection

// DefaultSignatures returns built-in rules that catch common attacks.
// These supplement any loaded rules file.
func DefaultSignatures() []Rule {
	return []Rule{
		// Port scans
		{ID: 1, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "TCP SYN Scan detected", Flags: "S", Severity: 2, Sid: 1000001},

		// NULL scan (no flags)
		{ID: 2, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "TCP NULL Scan detected", Flags: "0", Severity: 2, Sid: 1000002},

		// XMAS scan (FIN+URG+PSH)
		{ID: 3, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "TCP XMAS Scan detected", Flags: "FUP", Severity: 2, Sid: 1000003},

		// SSH brute force attempt
		{ID: 4, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "22",
			Msg: "Possible SSH brute force", Severity: 3, Sid: 1000004},

		// FTP brute force
		{ID: 5, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "21",
			Msg: "FTP connection attempt", Severity: 1, Sid: 1000005},

		// Telnet (plaintext)
		{ID: 6, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "23",
			Msg: "Telnet connection (cleartext protocol)", Severity: 2, Sid: 1000006},

		// SQL injection patterns
		{ID: 7, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "SQL injection attempt (UNION SELECT)", Content: "UNION SELECT", Nocase: true, Severity: 3, Sid: 1000007},

		{ID: 8, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "SQL injection attempt (OR 1=1)", Content: "OR 1=1", Nocase: true, Severity: 3, Sid: 1000008},

		// XSS
		{ID: 9, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "XSS attempt (<script>)", Content: "<script>", Nocase: true, Severity: 3, Sid: 1000009},

		// Shell command injection
		{ID: 10, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "Shell injection attempt (/bin/sh)", Content: "/bin/sh", Severity: 3, Sid: 1000010},

		// ICMP flood indicator
		{ID: 11, Action: "alert", Protocol: "icmp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "ICMP packet (potential flood)", Severity: 1, Sid: 1000011},

		// DNS suspicious TXT query
		{ID: 12, Action: "alert", Protocol: "udp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "53",
			Msg: "DNS TXT query (possible DNS tunneling)", Severity: 2, Sid: 1000012},

		// SMB / EternalBlue port
		{ID: 13, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "445",
			Msg: "SMB access (potential EternalBlue/ransomware)", Severity: 2, Sid: 1000013},

		// RDP brute force
		{ID: 14, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "3389",
			Msg: "RDP connection attempt", Severity: 2, Sid: 1000014},

		// Reverse shell signatures
		{ID: 15, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "Possible reverse shell (bash -i)", Content: "bash -i", Nocase: true, Severity: 3, Sid: 1000015},

		{ID: 16, Action: "alert", Protocol: "tcp", SrcIP: "any", SrcPort: "any",
			DstIP: "any", DstPort: "any",
			Msg: "Possible reverse shell (nc -e)", Content: "nc -e", Severity: 3, Sid: 1000016},
	}
}
